python/ql/lib/semmle/python/ApiGraphs.qll

```diff
@@ -474,6 +504,73 @@ module API {
     int getNumArgument() { result = count(this.getArg(_)) }
   }
 
+  /**
+   * A read of a subscript that is connected to the API graph.
+   *
+   * This can be used to reason about the index of subscripts.
+   * For the code
+   * ```python
+   * x = foo[bar]
+   * y = foo[baz]
+   * ```
+   * there wil be subscript edges from `foo` to both `foo[bar]` and `foo[baz]`,
+   * and there will be index edges from `foo` to both `bar` and `baz`.
+   * There will be a separate `SubscriptReadNode` for `foo[bar]` (and one for `foo[baz]`),
+   * and the member predicates `getObject` and `getIndex` will return nodes
+   * consistent with that particular read.
+   */
+  class SubscriptReadNode extends DataFlow::Node {
```

Comment:
Would you prefer this? For attributes we have them split up. And I would assume that models would mostly need a specific one, matching the modelled API..

Reply:
I'm mostly just wondering if it's _useful_ to have this distinction. It seems to me that the expression `a[b]`, regardless of whether it appears on the left hand side of an assignment or not, could still be thought of as an API source. 

Having things split up leads to a lot of duplication, e.g. the definition of `getSubscript/1` above. And even worse, by having two distinct classes, the return value of `getSubscript/1` can't be more specific than `API::Node`, which means you'll have to explicitly cast into the appropriate `Read`/`Write` class to use the convenience methods. (Contrast with `getACall() which returns an `API::CallNode`.)

To fix this, we could introduce a third `SubscriptRefNode` class (superclass of both `Read` and `Write`), but this would only help if you wanted to extract the object or index. For the value being written, you would still need to cast to `API::SubscriptWriteNode`.